File: ts/src/history.ts
/**
 * history — 选择记忆：记住每个选择器/歧义点上次选中的候选（history.toml）。
 *
 * 文件格式与 Go 引擎 BurntSushi/toml 输出完全兼容（同一数据目录共用）：
 *
 *	# picktui history — last selection per key (auto-managed)
 *	["git p"]
 *	last = "pull"
 *
 * 曾用名 picks.toml 仍会被自动迁移到新位置（与引擎行为一致）。
 * 文件不存在或损坏 → 空记录（记忆只是锦上添花，静默容错）。
 */
import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';

import { dataDir } from './config.js';
import { parseToml, escapeTomlString, keyLiteral } from './toml.js';
import { PicktuiError } from './types.js';

/** 选择记忆文件名；曾用名 picks.toml 因易与 pick 命令混淆而弃用。 */
export const HISTORY_FILE = 'history.toml';

/** 旧版文件名，仅用于首次读取时自动迁移。 */
export const LEGACY_HISTORY_FILE = 'picks.toml';

/** 注释头（对齐引擎输出，便于用户辨识文件来源）。 */
const HEADER = '# picktui history — last selection per key (auto-managed)';

/** 返回选择记忆文件路径。 */
export function historyPath(): string {
  return join(dataDir(), HISTORY_FILE);
}

/** history.toml 原始内容（label → { last }），与引擎同一份。 */
export function loadPickHistory(): Map<string, string> {
  const doc = loadHistoryToml();
  const out = new Map<string, string>();
  for (const label of doc.keys()) {
    const table = doc.get(label);
    if (!table) {
      continue;
    }
    const last = table.get('last');
    if (typeof last === 'string') {
      out.set(label, last);
    }
  }
  return out;
}

/** 读取/迁移 history.toml：不存在时静默返回空 Map。 */
function loadHistoryToml(): Map<string, Map<string, string | string[]>> {
  const p = historyPath();
  if (existsSync(p)) {
    return parseHistoryFile(readFileSync(p, 'utf8'));
  }
  const legacy = join(dataDir(), LEGACY_HISTORY_FILE);
  if (existsSync(legacy)) {
    const raw = parseHistoryFile(readFileSync(legacy, 'utf8'));
    try {
      writeHistoryEntries(raw);
      rmSync(legacy, { force: true });
    } catch {
      // 迁移写入失败不阻断读取——旧数据仍在内存中可用
    }
    return raw;
  }
  return new Map();
}

/** 解析 history.toml 文本；损坏时静默返回空（对齐引擎 LoadPickHistory）。 */
function parseHistoryFile(text: string): Map<string, Map<string, string | string[]>> {
  const doc = parseToml(text);
  if (doc === null) {
    return new Map();
  }
  return doc;
}

/** 取回 label 上次选中的值；无记录返回 ""。 */
export function lastPick(raw: Map<string, string>, label: string): string {
  return raw.get(label) ?? '';
}

/** 记录 label 的上次选中值（保留其它 label 的记忆后写回）。 */
export function savePick(label: string, value: string): void {
  const raw = loadPickHistory();
  raw.set(label, value);
  writeHistoryEntries(raw);
}

/** 将记忆写回 history.toml（自动建目录，对齐引擎输出格式）。 */
function writeHistoryEntries(
  entries: Map<string, Map<string, string | string[]>> | Map<string, string>,
): void {
  const lines: string[] = [HEADER];
  const labels = [...entries.keys()].sort();
  for (const label of labels) {
    lines.push(`[${keyLiteral(label)}]`);
    const table = entries.get(label);
    const last = table instanceof Map ? (table.get('last') as string | undefined) : table;
    lines.push(`last = "${escapeTomlString(last ?? '')}"`);
  }
  const p = historyPath();
  try {
    mkdirSync(dirname(p), { recursive: true });
    writeFileSync(p, lines.join('\n') + '\n', { mode: 0o644 });
  } catch (err) {
    throw new PicktuiError(
      `写入选择记忆失败：${err instanceof Error ? err.message : String(err)}`,
      1,
      '',
    );
  }
}

/** 读取 label 上次选中的值；无记录返回 ""。 */
export async function histGet(label: string): Promise<string> {
  return lastPick(loadPickHistory(), label);
}

/** 记录 label 的上次选中值（失败抛 PicktuiError，含退出码 1）。 */
export async function histSet(label: string, value: string): Promise<void> {
  savePick(label, value);
}

File: go/history.go
// history — 选择记忆：记住每个选择器/歧义点上次选中的候选，
// 下次弹窗时光标默认停在该候选上（--label 即记忆键）。
//
// 文件格式（TOML，位于数据目录 history.toml）：
//
//	# pickui history — last selection per key (auto-managed)
//	["git p"]
//	last = "pull"
//
// 纯文件存储、无 UI，不依赖 TUI，可独立单测。
package pickui

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"

	"github.com/BurntSushi/toml"
)

// HistoryFileName 选择记忆文件名，位于数据目录（DataDir()）。
// 曾用名 picks.toml 因与 `pick` 命令（动态候选过滤选择器）混淆而弃用。
const HistoryFileName = "history.toml"

// LegacyPickHistoryFileName 旧版选择记忆文件名，仅用于首次读取时自动迁移。
const LegacyPickHistoryFileName = "picks.toml"

// PickHistoryPath 返回选择记忆文件路径。
func PickHistoryPath() string {
	return filepath.Join(DataDir(), HistoryFileName)
}

// LoadPickHistory 解码 history.toml 为原始 map（label → {last: value}）。
// 新位置不存在但旧版 picks.toml 存在时自动迁移（读取后写入新位置并删除旧文件）；
// 文件不存在或损坏时返回空 map（静默忽略错误，记忆只是锦上添花）。
func LoadPickHistory() map[string]interface{} {
	raw, _ := LoadPickHistoryWithErr()
	if raw == nil {
		return map[string]interface{}{}
	}
	return raw
}

// LoadPickHistoryWithErr 与 LoadPickHistory 相同，但返回解析错误供调用方诊断。
// 文件不存在时返回 (空map, nil)；文件损坏时返回 (空map, error)。
func LoadPickHistoryWithErr() (map[string]interface{}, error) {
	raw := map[string]interface{}{}
	if _, err := os.Stat(PickHistoryPath()); err == nil {
		if _, err := toml.DecodeFile(PickHistoryPath(), &raw); err != nil {
			return map[string]interface{}{}, fmt.Errorf("解析 %s 失败: %w", PickHistoryPath(), err)
		}
		return raw, nil
	}
	legacy := filepath.Join(DataDir(), LegacyPickHistoryFileName)
	if _, err := os.Stat(legacy); err == nil {
		if _, err := toml.DecodeFile(legacy, &raw); err != nil {
			return map[string]interface{}{}, fmt.Errorf("解析旧版 %s 失败: %w", legacy, err)
		}
		if err := writeHistory(raw); err != nil {
			// 迁移写入失败不阻断读取——旧数据仍在内存中可用
			fmt.Fprintf(os.Stderr, "%s: 迁移选择记忆到 %s 失败: %v\n", Name(), PickHistoryPath(), err)
		} else {
			_ = os.Remove(legacy)
		}
		return raw, nil
	}
	return map[string]interface{}{}, nil
}

// LastPick 从记忆中取出 label 上次选中的值；无记录返回 ""。
func LastPick(raw map[string]interface{}, label string) string {
	if tbl, ok := raw[label].(map[string]interface{}); ok {
		if last, ok := tbl["last"].(string); ok {
			return last
		}
	}
	return ""
}

// SavePick 记录 label 的上次选中值，保留其它 label 的记忆后写回。
// 返回写入错误（如磁盘满、权限不足），调用方应记录但不必中断主流程——
// 选择记忆只是锦上添花的功能。
func SavePick(label, value string) error {
	raw := LoadPickHistory()
	raw[label] = map[string]interface{}{"last": value}
	return writeHistory(raw)
}

// writeHistory 将记忆写回 history.toml（自动建目录）。
func writeHistory(raw map[string]interface{}) error {
	var buf bytes.Buffer
	buf.WriteString(fmt.Sprintf("# %s history — last selection per key (auto-managed)\n", Name()))
	if err := toml.NewEncoder(&buf).Encode(raw); err != nil {
		return fmt.Errorf("编码 history.toml 失败: %w", err)
	}
	p := PickHistoryPath()
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return fmt.Errorf("创建目录 %s 失败: %w", filepath.Dir(p), err)
	}
	if err := os.WriteFile(p, buf.Bytes(), 0o644); err != nil {
		return fmt.Errorf("写入 %s 失败: %w", p, err)
	}
	return nil
}
